File: 3501-delete-nodes-from-linked-list-present-in-array/3501-delete-nodes-from-linked-list-present-in-array.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        Set<Integer> set= new HashSet<>();
        for(int i:nums)set.add(i);

        ListNode dummy= new ListNode();
        dummy.next=head;
         
         ListNode prev=dummy;

         while(head!=null){
            if(set.contains(head.val)){
                // if(head.next==null){
                //     head=null;
                //     head=head.next;
                // }
                prev.next=prev.next.next;
                head=prev.next;
            }
            else{
                prev=head;
            head=head.next;
            }
         }
         return dummy.next;

    }
}

File: 3501-delete-nodes-from-linked-list-present-in-array/3501-delete-nodes-from-linked-list-present-in-array.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        ListNode* prev = NULL;
        ListNode* after = NULL;
        ListNode* curr = head;
        unordered_set<int> st;
        for(int i:nums)st.insert(i);

        while(curr != NULL){
            if(st.find(curr->val) != st.end()){
                after = curr->next;
                if(prev != NULL)prev->next = after;
                else head = after;
            }else{
                prev = curr;
            }
            curr = curr->next;
        }
        return head;
    }
};